### ISSUE TITLE
Constraints result in dependencies being added instead of constraints

### ISSUE THREAD

**Issue by melix** [gradle/kotlin-dsl#710]
When I write the following code:

```
dependencies {
    constraints {
        "implementation"("com.google.collections:google-collections") {
            version {
                rejectAll()
            }
            because("Google collections is superceded by Guava")
        }
    }
}
```

The expectation is that what is added is a _constraint_. However, it turns out to be adding a _dependency_ instead. If I use the very same file in Groovy, a constraint is added.

### Expected Behavior

Extensions methods should be aware of the `constraints` block.

### Current Behavior

Constraints are created as hard dependencies, breaking the dependency graph.

**Comment by eskatos** [comment]
Workaround:
```kotlin
dependencies {
    constraints {
        add("implementation", "com.google.collections:google-collections") {
            version {
                rejectAll()
            }
            because("Google collections is superceded by Guava")
        }
    }
}
```

**Comment by JLLeitschuh** [comment]
@eskatos Do you have some grand solution for this problem in general or is it just to add the API in Kotlin every time issues like this arise in the API.

**Comment by eskatos** [comment]
@JLLeitschuh no, no grand solution atm 😄  Ad-hoc API additions is the easy way forward. But putting some thinking on this first sounds good. If something comes to you, please share.

**Comment by JLLeitschuh** [comment]
Annotating the Gradle core API and doing code generation at DSL build time popped into my head.

**Comment by ljacomet** [comment]
Just got bitten by this. The worst bit IMO is that it works but does something really different.
Is that something planned for a future (soon) release?

**Comment by melix** [comment]
I see that this issue is still open. It's a pretty serious bug, is this planned for 1.0?
